core/src/apps/cardano/address.py

```diff
@@ -1,137 +1,249 @@
-            [cbor.Tagged(24, address_data_encoded), crc.crc32(address_data_encoded)]
-        )
-    )
+BECH32_ADDRESS_PREFIX = "addr"
 
 
-def derive_address_and_node(
-    keychain: seed.Keychain, path: list, protocol_magic: int
-) -> Tuple[str, bip32.Node]:
-    node = keychain.derive(path)
+def validate_full_path(path: List[int]) -> bool:
+    """
+    Validates derivation path to fit {44', 1852'}/1815'/a'/{0,1,2}/i,
+    where `a` is an account number and i an address index.
+    The max value for `a` is 20, 1 000 000 for `i`.
+    """
+    if len(path) != 5:
+        return False
+    if path[0] != BYRON_PURPOSE and path[0] != SHELLEY_PURPOSE:
+        return False
+    if path[1] != 1815 | HARDENED:
+        return False
+    if path[2] < HARDENED or path[2] > 20 | HARDENED:
+        return False
+    if path[3] != 0 and path[3] != 1 and path[3] != 2:
+        return False
+    if path[4] > 1000000:
+        return False
+    return True
+
+
+def matches_with_network_id(address: bytes, network_id: int) -> bool:
+    return (address[0] & 0x0F) == network_id
 
-    # todo: GK - verify this is indeed correct (after IOHK confirmation/testnet support)
-    # protocol magic is included in Byron addresses only on testnets
-    if protocol_magic == ProtocolMagics.MAINNET:
-        address_attributes = {}
-        address_payload = None
+
+def has_byron_address_header(address: bytes) -> bool:
+    return (address[0] >> 4) == AddressId.BYRON_ADDRESS
+
+
+def get_human_readable_address(address: bytes, is_byron_address: bool) -> str:
+    if is_byron_address:
+        return base58.encode(address)
     else:
-        address_attributes = {2: cbor.encode(protocol_magic)}
-        address_payload = [address_attributes]
+        return bech32_encode(BECH32_ADDRESS_PREFIX, address)
 
-    address_root = _get_address_root(node, address_payload)
-    address_type = 0
-    address_data = [address_root, address_attributes, address_type]
-    address_data_encoded = cbor.encode(address_data)
 
-    return (_encode_address_raw(address_data_encoded), node)
+def get_public_key_hash(keychain: seed.Keychain, path: List[int]) -> bytes:
+    node = keychain.derive(path)
+    public_key = remove_ed25519_prefix(node.public_key())
+    return hashlib.blake2b(data=public_key, outlen=28).digest()
 
 
-def is_safe_output_address(address) -> bool:
-    """
-    Determines whether it is safe to include the address as-is as
-    a tx output, preventing unintended side effects (e.g. CBOR injection)
-    """
-    address_data_encoded = _get_address_data(address)
-    if address_data_encoded is None:
-        return False
+def path_to_staking_path(path: List[int]) -> List[int]:
+    return path[:3] + [2, 0]
 
-    return _encode_address_raw(address_data_encoded) == address
 
+def derive_address(
+    keychain: seed.Keychain,
+    parameters: CardanoAddressParametersType,
+    protocol_magic: int,
+    network_id: int,
+    human_readable: bool = True,
+) -> str:
```

Comment:
this actually returns bytes or string - IMHO it would be cleaner to have separate `derive_address_bytes` and `derive_human_readable address` - if you currently want a human readable address you are setting the named parameter human_readable anyway and if you want the bytes you call the function without any parameter which is kind of misleading because the natural expectation from a function called `derive_address` is that you get an address in it's "natural", i.e. human readable form, so it may be confusing

Reply:
Split into two functions.